### ISSUE TITLE
Improve time step calculation performance (not first priority)

### ISSUE THREAD

**Issue by xaviernogueira** [EcohydrologyTeam/ClearWater-modules#24]
Once we get closer to having sub-modules completed, we can focus on improving the performance of the core `base.Model` code. Since this code is inherited, any benefits are passed to all sub-modules. Some experimenting can/will be done to see if we can squeeze out time_step performance. Model __init__ performance is less relevant.

We can use `snakeviz` with `Cprofiler` to assess any computational rate limiting steps.

**Comment by xaviernogueira** [comment]
**Update -> some promising results!**

I took the time to do some experimentation on the core "computation engine", this is currently being done on a notebook in 
 a separate branch but will be merged in. By "computation engine" I am referring to the part of the code that applies all calculations, regardless of which module is being used.

So far I tested:
* V1 - The existing set up, where we use `xr.apply_ufunc` in a function that is JIT compiled with `@numba.jit(forceobj=True)`. 
* V2 - The same as V1, except no more JIT compiling. It is well documented that `forceobj=True` dramatically reduces the efficacy of JIT.
* V3 - A new approach where we use `map()` to create an iterable that contains everything necessary to update a timestep, but as numpy arrarys NOT `xr.DataArrays`.
* V4 - An attemt to Numba JIT the inner loop of V3 -> turns out this is impossible, as each set of inputs will have a different process function, and Numba `nopython=True` mode can't handle "heterogenous list" inputs.

**Results:**
As shown below, we can see that the new V3 approach is nearly 50% faster than the existing method (17ms shaved off). V2 did not speed things up at all. We can also see that the `iter_computation()` call within `increment_timestep` is responsible for 76% of the run time. Therefore our 50% reduction in `iter_computation()`` should account for a 36% reduction in timestep run time.
![image](https://github.com/EcohydrologyTeam/ClearWater-modules/assets/58796351/e6b8e4bb-46b9-41fd-bdac-32e3e517b0d4)

**Next steps:**
I am going to implement V3 into the main modules, see if all tests pass, and if so I will merge into the main branch.
